### PR TITLE
Explain cron start time skew

### PR DESCRIPTION
**Pull request by mattmanning**
## Release Playbook
- [ ] Rebase against master
- [ ] Code review
- [ ] Merge into master
- [ ] Verify changes at http://site-staging.convox.com
- [ ] Promote release on `site-production` in the `convox/production` Rack

### CODE REVIEW

**Comment by mattmanning** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;knWwmUgr&#x2F;692-explain-cron-start-time-skew"> Explain cron start time skew</a>

File: _deployment/scheduled-tasks.md [hunk 1]
```diff
@@ -40,7 +40,9 @@ Cron expressions use the following format:
 ```
 
 <div class="block-callout block-show-callout type-info" markdown="1">
-  Using both day-of-week and day-of-month in the same expression is not supported. One of these fields must be a `?`.
+  Using both day-of-week and day-of-month in the same expression is not supported. One of these fields must be a `?`.<br/>
+  <br/>
+  The actual start time of a job may be skewed up to 10 seconds. This is done to prevent infrastructure throttling errors when lots of jobs start at the same time.
```
Comment (ianmalott) [inline _deployment/scheduled-tasks.md]: That seems sufficient to reassure users who are seeing unexpected skew, but out of curiosity, what exactly makes a job a target for skew? Is it just having a time that matches another job's time down to seconds?

This is a bit of a stretch, but let's say I wanted to schedule a few jobs to run at roughly the same time, but in a particular order, how would I do that while making sure skew did not enter into the picture?

Comment (mattmanning) [inline _deployment/scheduled-tasks.md]: @ianmalott All jobs are skewed now, so I don't think any are particularly targeted. I also added some language to answer your ordering question.

**Comment by mattmanning** [comment]
@ianmalott Here's the change where the cron skew was added: https://github.com/convox/rack/pull/1205/files

If you look at the changes to `api/models/templates/app.tmpl`, lines 561 and 564 are relevant. The rest is adding logging.

You can see that a random number between 0 and 10000 (representing milliseconds) is generated and then that is used in a call to `setTimeout`, randomly delaying the `ecs.runTask` by up to 10 seconds. This helps to guarantee that if there are several tasks scheduled to run at the same time, we don't make a bunch of `ecs` calls at once and get throttled.

**Comment by ianmalott** [comment]
@mattmanning thanks for the explanation. That helps a lot.

**Pull request merged**